web/webkit/src/main/scala/net/liftweb/http/LAFutureWithSession.scala

```diff
@@ -0,0 +1,48 @@
+package net.liftweb.http
+
+import net.liftweb.actor.{LAFuture, LAScheduler}
+import net.liftweb.common.{EmptyBox, Failure, Full}
+
+
+object LAFutureWithSession {
+
+  /**
+    * Creates `LAFuture` instance aware of current request and session. Each `LAFuture` returned by chained
+    * transformation method (e.g. `map`, `flatMap`) will be also request/session-aware. However, it's
+    * important to bear in mind that initial session or request are not propagated to chained methods. It's required
+    * that current execution thread for chained method has its own request or session available if reading/writing
+    * some data to it as a part of chained method execution.
+    */
+  def withCurrentSession[T](task: => T, scheduler: LAScheduler = LAScheduler): LAFuture[T] = {
+    S.session match {
+      case Full(session) =>
+        withSession(task, scheduler)
+
+      case empty: EmptyBox =>
+        withFailure(empty ?~! "LiftSession not available in this thread context", scheduler)
```

Comment:
I'd say that *sometimes* we throw an exception if method needs a session but it's not available, just like `S ?` do, but when *something like this attempts to access the session without a proper session context* we usually return a `Failure`. This is also similar for `Req`. 

If you take a look at `S`, you'll notice that `S.session` returns a `Box`. As a result, almost all functions requiring session to be available return `Box` as well (e.g. `findOrCreateComet`, `getSessionAttribute`, `runTemplate`, `eval`). Functions returning `Boolean` in `S` usually return `false` if session is not available (e.g. `legacyIeCompatibilityMode`). I'd rather say that `S.?` is an exception here.

Similar situation is for `S.request`. It returns `Box` and as a result other methods requiring request to be available return `Box` as well (`uri`, `queryString`, `getRequestHeader`, `referer`).